docs/adr/0001-pipeline-to-publish-python-packages.md

```diff
@@ -0,0 +1,36 @@
+# Pipeline to publish Python packages
+
+## Status
+
+Proposed.
+
+## Context
+
+Based on my experience over the last year using Python in a data science context, functionality typically begins life in a Jupyter notebook.  There then comes a point where some of that functionality will then be:
+
+1. Promoted into a production environment, and/or;
+1. Re-used across multiple projects within Jupyter to provide tooling in areas such as ingestion, wrangling / feature engineering or exploration / analysis.
+
+There then comes the need to "productionise" the code - for example:
+
+1. Encapsulation of code in appropriate project / class structure with a view to building core set of re-useable fucntionality;
+1. Dependency management - including managing packages that are "dev only";
+1. Automated unit tests / test coverage analysis;
+1. Linting;
+1. Publlishing as a Python package that can imported in multiple use cases : Jupter notebooks, Spark jobs and other operational processes as a first class citizen alongside 3rd party Python packages.
+
+## Decision
+
+Develop a best practice Azure DevOps pipeline (and associated documentation) that will enable a Python repository to be developed and deployed as an Azure DevOps Artefact.
```

Comment:
Will this be an exemplar, intended for clipboard reuse? Or will any of what we create in this repo be available through some sort of distribution mechanism that enables us to push out updates? (E.g., suppose we work out a way of improving something after a few months. Will it be possible to make those improvements available automatically to anyone using this, like you can with a NuGet package by publishing a new version, or will people need to manually realign with the current state of this repo if they want to benefit from improvements?)

Reply:
The proposal would be to use cookiecutter for the project template.  I'm not an expert but having scanned the documentation on GitHub, it doesn't seem like you are able to apply incremental improvementsinto projects created with prior versions of the template.

Reply:
With our existing build recommended practices, we've defined built templates that are then used from other projects, meaning that when we fix bugs in those templates, everything using them benefits. So it's possible you could do something similar. I'm not sufficiently familiar with the details of Python to know what else might be possible - is there anything we could do with Python's packaging system(s) so that projects using this system aren't just taking a snapshot?

Reply:
I imagine cookie-cutter would be used for the initial python ‘solution’ or project structure, albeit including the pipeline. So whilst the pipeline could be constructed similar to the way we do today (with shared templates that can be updated centrally), it wouldn’t touch the code base structure itself. However, neither does our existing tooling for .NET so far as I’ve seen.